File: StockAlerts.java
// StockAlerts.java
import java.io.*;
import java.nio.file.*;
import java.net.*;
import java.util.*;
import java.time.*;
import com.google.gson.*;

class Alert {
    String id;
    String symbol;
    double threshold;
    boolean above;
    boolean triggered;

    Alert() {}
    Alert(String symbol, double threshold, boolean above) {
        this.id = UUID.randomUUID().toString().substring(0,8);
        this.symbol = symbol.toUpperCase();
        this.threshold = threshold;
        this.above = above;
        this.triggered = false;
    }
}

public class StockAlerts {
    private List<Alert> alerts = new ArrayList<>();
    private final String configFile = "alerts.json";
    private final Gson gson = new GsonBuilder().setPrettyPrinting().create();
    private static final String API_KEY = "demo";
    private static final String BASE_URL = "https://www.alphavantage.co/query";

    public StockAlerts() { load(); }

    private void load() {
        try {
            Path path = Paths.get(configFile);
            if (Files.exists(path)) {
                String json = new String(Files.readAllBytes(path));
                Alert[] arr = gson.fromJson(json, Alert[].class);
                alerts = Arrays.asList(arr);
            }
        } catch (Exception e) {}
    }

    private void save() {
        try {
            Files.write(Paths.get(configFile), gson.toJson(alerts).getBytes());
        } catch (Exception e) {}
    }

    private Alert getAlert(String symbol) {
        for (Alert a : alerts) {
            if (a.symbol.equalsIgnoreCase(symbol)) return a;
        }
        return null;
    }

    public void add(String symbol, double threshold, boolean above) {
        Alert existing = getAlert(symbol);
        if (existing != null) alerts.remove(existing);
        Alert alert = new Alert(symbol, threshold, above);
        alerts.add(alert);
        save();
        System.out.printf("✅ Added alert: %s %s %.2f%n", symbol, above ? ">" : "<", threshold);
    }

    public void remove(String symbol) {
        Alert existing = getAlert(symbol);
        if (existing != null) {
            alerts.remove(existing);
            save();
            System.out.printf("✅ Removed alert for %s%n", symbol.toUpperCase());
        } else {
            System.out.printf("No alert for %s%n", symbol.toUpperCase());
        }
    }

    public void list() {
        if (alerts.isEmpty()) {
            System.out.println("No alerts.");
            return;
        }
        System.out.println("\n📋 Current alerts:");
        for (Alert a : alerts) {
            String status = a.triggered ? "🔔" : "⏳";
            System.out.printf("  %s: %s %.2f %s%n", a.symbol, a.above ? ">" : "<", a.threshold, status);
        }
    }

    private Double fetchPrice(String symbol) {
        try {
            String url = BASE_URL + "?function=GLOBAL_QUOTE&symbol=" + symbol + "&apikey=" + API_KEY;
            HttpURLConnection conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setRequestMethod("GET");
            BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
            reader.close();
            JsonObject obj = gson.fromJson(sb.toString(), JsonObject.class);
            JsonObject quote = obj.getAsJsonObject("Global Quote");
            if (quote == null) return null;
            return quote.get("05. price").getAsDouble();
        } catch (Exception e) {
            System.err.println("Error fetching " + symbol + ": " + e.getMessage());
            return null;
        }
    }

    public void checkAlerts() {
        for (Alert a : alerts) {
            Double price = fetchPrice(a.symbol);
            if (price == null) continue;
            boolean triggered = a.above ? price > a.threshold : price < a.threshold;
            if (triggered && !a.triggered) {
                a.triggered = true;
                String direction = a.above ? "above" : "below";
                System.out.printf("\n🔔 ALERT: %s price is %.2f (%s %.2f)!%n", a.symbol, price, direction, a.threshold);
                save();
            }
        }
    }

    public void watch(int interval) throws InterruptedException {
        System.out.printf("📈 Monitoring alerts every %ds (press Ctrl+C to stop)%n", interval);
        while (true) {
            checkAlerts();
            Thread.sleep(interval * 1000L);
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.out.println("Usage: StockAlerts <command> [options]");
            return;
        }
        StockAlerts app = new StockAlerts();
        String cmd = args[0];

        switch (cmd) {
            case "add": {
                if (args.length < 3) { System.out.println("add <symbol> <threshold> [--above|--below]"); return; }
                String symbol = args[1];
                double threshold = Double.parseDouble(args[2]);
                boolean above = true;
                for (int i=3; i<args.length; i++) {
                    if (args[i].equals("--below")) above = false;
                }
                app.add(symbol, threshold, above);
                break;
            }
            case "remove": {
                if (args.length < 2) { System.out.println("remove <symbol>"); return; }
                app.remove(args[1]);
                break;
            }
            case "list":
                app.list();
                break;
            case "watch": {
                int interval = 60;
                for (int i=1; i<args.length; i++) {
                    if (args[i].equals("--interval") && i+1 < args.length) {
                        interval = Integer.parseInt(args[++i]);
                    }
                }
                app.watch(interval);
                break;
            }
            default:
                System.out.println("Unknown command.");
        }
    }
}

File: stock_alerts.js
// stock_alerts.js
#!/usr/bin/env node
const fs = require('fs');
const axios = require('axios');
const { program } = require('commander');
const { v4: uuidv4 } = require('uuid');

const CONFIG_FILE = 'alerts.json';
const API_KEY = 'demo';
const BASE_URL = 'https://www.alphavantage.co/query';

class Alert {
    constructor(symbol, threshold, above = true, id = null) {
        this.id = id || uuidv4().slice(0,8);
        this.symbol = symbol.toUpperCase();
        this.threshold = threshold;
        this.above = above;
        this.triggered = false;
    }

    check(price) {
        const triggered = this.above ? price > this.threshold : price < this.threshold;
        if (triggered && !this.triggered) {
            this.triggered = true;
            return true;
        }
        return false;
    }
}

class App {
    constructor() {
        this.alerts = [];
        this.load();
    }

    load() {
        if (fs.existsSync(CONFIG_FILE)) {
            try {
                const data = JSON.parse(fs.readFileSync(CONFIG_FILE));
                this.alerts = data.map(a => {
                    const alert = new Alert(a.symbol, a.threshold, a.above, a.id);
                    alert.triggered = a.triggered || false;
                    return alert;
                });
            } catch (e) {}
        }
    }

    save() {
        fs.writeFileSync(CONFIG_FILE, JSON.stringify(this.alerts.map(a => ({
            id: a.id,
            symbol: a.symbol,
            threshold: a.threshold,
            above: a.above,
            triggered: a.triggered
        })), null, 2));
    }

    add(symbol, threshold, above = true) {
        // Remove existing
        this.alerts = this.alerts.filter(a => a.symbol !== symbol.toUpperCase());
        const alert = new Alert(symbol, threshold, above);
        this.alerts.push(alert);
        this.save();
        console.log(`✅ Added alert: ${symbol} ${above ? '>' : '<'} ${threshold.toFixed(2)}`);
    }

    remove(symbol) {
        this.alerts = this.alerts.filter(a => a.symbol !== symbol.toUpperCase());
        this.save();
        console.log(`✅ Removed alert for ${symbol.toUpperCase()}`);
    }

    list() {
        if (this.alerts.length === 0) {
            console.log('No alerts.');
            return;
        }
        console.log('\n📋 Current alerts:');
        for (const a of this.alerts) {
            const status = a.triggered ? '🔔' : '⏳';
            console.log(`  ${a.symbol}: ${a.above ? '>' : '<'} ${a.threshold.toFixed(2)} ${status}`);
        }
    }

    async fetchPrice(symbol) {
        const url = `${BASE_URL}?function=GLOBAL_QUOTE&symbol=${symbol}&apikey=${API_KEY}`;
        try {
            const resp = await axios.get(url, { timeout: 10000 });
            const quote = resp.data['Global Quote'];
            if (!quote) return null;
            return parseFloat(quote['05. price']);
        } catch (e) {
            console.error(`Error fetching ${symbol}: ${e.message}`);
            return null;
        }
    }

    async checkAlerts() {
        for (const alert of this.alerts) {
            const price = await this.fetchPrice(alert.symbol);
            if (price === null) continue;
            if (alert.check(price)) {
                const direction = alert.above ? 'above' : 'below';
                console.log(`\n🔔 ALERT: ${alert.symbol} price is ${price.toFixed(2)} (${direction} ${alert.threshold.toFixed(2)})!`);
                this.save();
            }
        }
    }

    async watch(interval = 60) {
        console.log(`📈 Monitoring alerts every ${interval}s (press Ctrl+C to stop)`);
        while (true) {
            await this.checkAlerts();
            await new Promise(resolve => setTimeout(resolve, interval * 1000));
        }
    }
}

program
    .command('add <symbol> <threshold>')
    .option('--above', 'Alert when price goes above threshold', true)
    .option('--below', 'Alert when price goes below threshold')
    .action((symbol, threshold, options) => {
        const app = new App();
        const above = !options.below;
        app.add(symbol, parseFloat(threshold), above);
    });

program
    .command('remove <symbol>')
    .action((symbol) => {
        const app = new App();
        app.remove(symbol);
    });

program
    .command('list')
    .action(() => {
        const app = new App();
        app.list();
    });

program
    .command('watch')
    .option('--interval <n>', 'Check interval in seconds', parseInt, 60)
    .action(async (options) => {
        const app = new App();
        await app.watch(options.interval);
    });

program.parse(process.argv);
